### ISSUE TITLE
DeviceCodeCredential() error: "need admin approval"

### ISSUE THREAD

**Issue by rfernand2** [Azure/azure-sdk-for-python#10155]
azure-identity==1.2.0
azure-keyvault-secrets==4.0.1
Windows 10
python 3.6.10

My command line app can successfully use InteractiveBrowserCredential() to get credentials for the user when the OS has a browser, but for SSH connections, we need a solution like DeviceCodeCredential().  For some reason, DeviceCodeCredential requires a client id, and is difficult to find the right configuration to make it work.  Why can't it be as easy as InteractiveBrowserCredential() to use, but just provide an alternative UI?  

Anyway, I created an app with AAD, gave it permission to access the Azure Key Vault (user impersonation).  Using the client id of this app (is that the client id I am supposed to use??), it prompts the user to sign in, give permission to the app, and then issues the error "Need Admin Approval".  Huh? 

I'm not sure if its asking referring to the admin for the app (which is me, I think), the admin for my service subscription (which is me, I think), or possible an admin for my company (Microsoft) for all Azure accounts.  If its is not me, I have no idea how to find this admin person.  And should that even be necessary?  All of this is not needed for the InteractiveBrowserCredential() API.  

BTW, when I added the API permission to access the Key Vault to my app, it said that Admin Approval was **not** required.  

I must be missing something - can anyone provide some guidance here?  Thanks in advance.

**Describe the bug**

**To Reproduce**
1. 

**Expected behavior**

**Screenshots**

**Additional context**

**Comment by v-xuto** [comment]
Hi @rfernand2, Here is a [sample](https://github.com/Azure/azure-sdk-for-python/blob/bb62c962f4b145e0363e7dad7bf277a9027167a5/sdk/identity/azure-identity/tests/test_live.py#L16) of DeviceCodeCredential. 
For the value of the parameter cliend_id in DeviceCodeCredential, you should use AZURE_CLI_CLIENT_ID directly in the azure/identity/_constants.py file.  
  
Sample code snippet:   
```
from azure.identity import DeviceCodeCredential 
from azure.identity._constants import AZURE_CLI_CLIENT_ID 
from azure.keyvault.secrets import SecretClient 
  
credential = DeviceCodeCredential(AZURE_CLI_CLIENT_ID) 
client = SecretClient(VAULT_URL, credential) 
secret = client.set_secret("secret", "the test passed") 
assert secret.value == "the test passed" 
 ```
Please feel free to reach us with more detailed information if the error still persists by using the sample. 
@jongio

**Comment by rfernand2** [comment]
Hi @v-xuto, **thanks** for the code sample.  I love the simplicity - who knew there was an AZURE_CLI_CLIENT_ID?!  Any way, I just pasted in my Vault URL and tried it.  I get the following error:

> (exper) c:\github\ExperimentTools\cmdlineTest>python azure_device_code_test.py
> To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code BFNHY86BX to authenticate.
> Traceback (most recent call last):
>   File "azure_device_code_test.py", line 9, in <module>
>     secret = client.set_secret("secret", "the test passed")
>   File "c:\anaconda3\envs\exper\lib\site-packages\azure\core\tracing\decorator.py", line 71, in wrapper_use_tracer
>     return func(*args, **kwargs)
>   File "c:\anaconda3\envs\exper\lib\site-packages\azure\keyvault\secrets\_client.py", line 117, in set_secret
>     **kwargs
>   File "c:\anaconda3\envs\exper\lib\site-packages\azure\keyvault\secrets\_shared\_generated\v7_0\operations\_key_vault_client_operations.py", line 1434, in set_secret
>     map_error(status_code=response.status_code, response=response, error_map=error_map)
>   File "c:\anaconda3\envs\exper\lib\site-packages\azure\core\exceptions.py", line 83, in map_error
>     raise error
> azure.core.exceptions.HttpResponseError: (Forbidden) Access denied to first party service.
> Caller: name=from-infra;tid=f8cdef31-a31e-4b4a-93e4-5f571e91255a;appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;iss=https://sts.windows.net/f8cdef31-a31e-4b4a-93e4-5f571e91255a/
> Vault: xtsandboxvault;location=eastus

Maybe I am using a different version than you?  My info:

> azure-identity==1.2.0
> azure-keyvault-secrets==4.0.1
> Windows 10
> python 3.6.10

**Comment by v-xuto** [comment]
Hi @rfernand2,  
When you run azure_device_code_test.py file, you need according to the following instructions on the browser to open the link https://microsoft.com/devicelogin, then enter the corresponding code (such as BFNHY86BX) for authentication. I wonder if you have done so? 
``` 
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code BFNHY86BX to authenticate. 
```

**Comment by rfernand2** [comment]
Yes, for the above test, I was running on my home computer (tried it multiple times).  The web page gave me the response "You have signed in to the Microsoft Azure Cross-platform Command Line Interface application on your device. You may now close this window.".

However, I just ran the same test at work, and... **it worked!**  Same confirmation message from the browser but without the stack trace, and I verified that I can now read and write secrets to the vault.  This was on Windows; I also verified its works from my Linux machine.  This is **awesome**.  

Before we close this issue, let me try again from my home PC.  Any reason why it wouldn't work there?

**Comment by rfernand2** [comment]
Also, can you direct me to any documentation you have on **AZURE_CLI_CLIENT_ID**?  I can't seem to find much about it searching the web.  I am trying to figure out if there are condition under which it won't work.  Thanks...

**Comment by chlowell** [comment]
> I'm not sure if its asking referring to the admin for the app (which is me, I think), the admin for my service subscription (which is me, I think), or possible an admin for my company (Microsoft)
>

It's the third one, an admin for your Azure AD tenant needs to grant the permission to your application.

> Before we close this issue, let me try again from my home PC. Any reason why it wouldn't work there?
> 
Unless you use a VPN (or live at the office), you're on a different network at home. Your Key Vault may have network access restrictions ([docs](https://docs.microsoft.com/en-us/azure/key-vault/key-vault-network-security)).

>Also, can you direct me to any documentation you have on AZURE_CLI_CLIENT_ID?  I am trying to figure out if there are any conditions under which it won't work

It's undocumented. The `_constants` module is internal and changes without warning; this constant in particular will be removed in a future release. The best practice is to have your application registered with Azure AD, with the permissions it requires.

**Comment by chlowell** [comment]
Thanks for opening this issue, please open another if you encounter further problems. I'm closing this issue because the mentioned problems are due to Azure AD and Key Vault configuration, and can't be resolved by the SDK.
